waspc/src/Util/Control/Monad.hs

```diff
@@ -0,0 +1,40 @@
+module Util.Control.Monad
+  ( foldMapM',
+    foldM1,
+  )
+where
+
+import Control.Monad (foldM)
+import Data.List (foldl')
+import Data.List.NonEmpty (NonEmpty ((:|)))
+
+-- | Analogous to "Data.Foldable.foldMap'", except that its result is encapsulated in a
+-- monad.
+--
+-- @
+-- foldMapM f [x1, x2, ..., xn] ==
+--   do
+--     a1 <- f x1
+--     a2 <- f x2
+--     ...
+--     an <- f xn
+--     return $ mempty <> a1 <> a2 <> ... <> an
+-- @
+--
+-- __Examples__
+--
+-- >>> import Data.Monoid
+-- >>> getSum <$> foldMapM' (\n -> if n > 3 then Right n else Left n) (map Sum [4,7,5,6])
+-- Right 22
+foldMapM' :: (Foldable t, Monad m, Monoid s) => (a -> m s) -> t a -> m s
+foldMapM' f = foldl' (\ms a -> ms >>= \s -> (s <>) <$> f a) $ pure mempty
```

Comment:
Would it make sense to write some tests for this function, and the function below? I guess so? I don't think it is worthing wasting time on that now, but maybe just write a TODO.